src/upgrades/mod.rs

```diff
@@ -129,94 +260,175 @@ pub trait UpgradeStore {
+                Err(UpgradeError::custom(msg))
+            } else if versions.from < KrillVersion::release(0, 9, 0) {
+                // We need to prepare pubd first, because if there were any CAs using
+                // an embedded repository then they will need to be updated to use the
+                // RFC 8181 protocol (using localhost) instead, and this can only be
+                // *after* the publication server data is migrated.
+                PubdObjectsMigration::prepare(mode, config.clone())?;
+
+                // We fool a repository manager for the CA migration to use the upgrade
+                // data directory as its base dir. This repository manager will be used
+                // to get the repository response XML for any (if any) CAs that were
+                // using an embedded repository.
+                let mut repo_manager_migration_config = (*config).clone();
+                repo_manager_migration_config.data_dir = config.upgrade_data_dir();
+
+                // We need a signer because it's required by the repo manager, although
+                // we will not actually use it during the migration. Let it use the
+                // config using the upgrade_data_dir as base dir to ensure that it
+                // cannot - even unintendedly - affect any of they keys.
+                let signer = Arc::new(KrillSigner::build(&repo_manager_migration_config.data_dir)?);
+
+                let repo_manager = RepositoryManager::build(Arc::new(repo_manager_migration_config), signer)?;
+
+                CaObjectsMigration::prepare(mode, config, repo_manager)?;
+
+                Ok(Some(UpgradeReport::new(true, versions)))
+            } else {
+                Ok(Some(UpgradeReport::new(false, versions)))
+            }
         }
-
-        Ok(())
-    }
-
-    fn archive_to_migration_scope(&self, key: &KeyStoreKey) -> Result<(), UpgradeError> {
-        self.store()
-            .archive_to(key, MIGRATION_SCOPE)
-            .map_err(UpgradeError::KeyStoreError)
     }
+}
 
-    fn drop_migration_scope(&self, scope: &str) -> Result<(), UpgradeError> {
-        let scope = format!("{}/{}", scope, MIGRATION_SCOPE);
-        self.store().drop_scope(&scope).map_err(UpgradeError::KeyStoreError)
+/// Finalise the data migration for an upgrade. I.e. move the prepared data and archive
+/// the old data if applicable to this upgrade, and otherwise (in any event) update the
+/// the current versions for the "cas" and "pubd" store where applicable.
+pub fn finalise_data_migration(upgrade: &UpgradeVersions, config: &Config) -> UpgradeResult<()> {
+    // Move directories - if applicable (servers can have cas, repo server or both)
+
+    let from = upgrade.from();
+    let current = upgrade.to();
+    let data_dir = &config.data_dir;
+    let upgrade_dir = config.upgrade_data_dir();
+
+    // cas -> arch-cas-{old-version}
+    // upgrade-data/cas -> cas
+    // upgrade-data/ca_objects -> ca_objects
```

Comment:
So is there no `ca_objects` dir prior to the upgrade then?

Reply:
indeed